store/streaming/file/service.go

```diff
@@ -0,0 +1,279 @@
+	srcChan            <-chan []byte                          // the channel that all of the WriteListeners write their data out to
+	filePrefix         string                                 // optional prefix for each of the generated files
+	writeDir           string                                 // directory to write files into
+	codec              codec.BinaryCodec                      // marshaller used for re-marshalling the ABCI messages to write them out to the destination files
+	stateCache         [][]byte                               // cache the protobuf binary encoded StoreKVPairs in the order they are received
+	stateCacheLock     *sync.Mutex                            // mutex for the state cache
+	currentBlockNumber int64                                  // the current block number
+	currentTxIndex     int64                                  // the index of the current tx
+}
+
+// IntermediateWriter is used so that we do not need to update the underlying io.Writer inside the StoreKVPairWriteListener
+// everytime we begin writing to a new file
+type IntermediateWriter struct {
+	outChan chan<- []byte
+}
+
+// NewIntermediateWriter create an instance of an intermediateWriter that sends to the provided channel
+func NewIntermediateWriter(outChan chan<- []byte) *IntermediateWriter {
+	return &IntermediateWriter{
+		outChan: outChan,
+	}
+}
+
+// Write satisfies io.Writer
+func (iw *IntermediateWriter) Write(b []byte) (int, error) {
+	iw.outChan <- b
+	return len(b), nil
+}
+
+// NewStreamingService creates a new StreamingService for the provided writeDir, (optional) filePrefix, and storeKeys
+func NewStreamingService(writeDir, filePrefix string, storeKeys []sdk.StoreKey, c codec.BinaryCodec) (*StreamingService, error) {
+	listenChan := make(chan []byte)
+	iw := NewIntermediateWriter(listenChan)
+	listener := types.NewStoreKVPairWriteListener(iw, c)
+	listeners := make(map[sdk.StoreKey][]types.WriteListener, len(storeKeys))
+	// in this case, we are using the same listener for each Store
+	for _, key := range storeKeys {
+		listeners[key] = append(listeners[key], listener)
+	}
+	// check that the writeDir exists and is writeable so that we can catch the error here at initialization if it is not
+	// we don't open a dstFile until we receive our first ABCI message
+	if err := isDirWriteable(writeDir); err != nil {
+		return nil, err
+	}
+	return &StreamingService{
+		listeners:      listeners,
+		srcChan:        listenChan,
+		filePrefix:     filePrefix,
+		writeDir:       writeDir,
+		codec:          c,
+		stateCache:     make([][]byte, 0),
+		stateCacheLock: new(sync.Mutex),
+	}, nil
+}
+
+// Listeners returns the StreamingService's underlying WriteListeners, use for registering them with the BaseApp
+func (fss *StreamingService) Listeners() map[sdk.StoreKey][]types.WriteListener {
+	return fss.listeners
+}
+
+// ListenBeginBlock satisfies the Hook interface
```

Comment:
we should make sure we remove the usage of the Hook nomenclature in this file.

Reply:
100%, good catch